### PR TITLE
Gtk subclassing round 1

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Adds subclassing support to the following types
- `GtkButton`
- `GtkAdjustment`
- `GtkListBoxRow`
- `GtkFlowBoxChild`
- `GtkDrawingArea`

Removes GtkHeaderBar because it's not derivable anymore

Part of #56

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
The CI is failing due to https://github.com/actions-rs/toolchain/issues/124

**Comment by sdroege** [comment]
Did you check against / take all those from gtk3-rs?

File: gtk4/src/subclass/adjustment.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+impl<T: AdjustmentImpl> AdjustmentImplExt for T {
+    fn parent_changed(&self, adjustment: &Self::Type) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gtk_sys::GtkAdjustmentClass;
+            let f = (*parent_class)
+                .changed
+                .expect("No parent class impl for \"changed\"");
```
Comment (sdroege) [inline gtk4/src/subclass/adjustment.rs:32]: This and the other vfunc is optional and simply nothing should happen if it's `NULL`. See the C code.

File: gtk4/src/subclass/button.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+impl<T: ButtonImpl> ButtonImplExt for T {
+    fn parent_activate(&self, button: &Self::Type) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gtk_sys::GtkButtonClass;
+            let f = (*parent_class)
+                .activate
+                .expect("No parent class impl for \"activate\"");
```
Comment (sdroege) [inline gtk4/src/subclass/button.rs:33]: These two are also optional

File: gtk4/src/subclass/drawing_area.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+impl<T: DrawingAreaImpl> DrawingAreaImplExt for T {
+    fn parent_resize(&self, drawing_area: &Self::Type, width: i32, height: i32) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class =
+                data.as_ref().get_parent_class() as *mut gtk_sys::GtkDrawingAreaClass;
+            let f = (*parent_class)
+                .resize
+                .expect("No parent class impl for \"resize\"");
```
Comment (sdroege) [inline gtk4/src/subclass/drawing_area.rs:29]: Optional

File: gtk4/src/subclass/flow_box_child.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+impl<T: FlowBoxChildImpl> FlowBoxChildImplExt for T {
+    fn parent_activate(&self, child: &Self::Type) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class =
+                data.as_ref().get_parent_class() as *mut gtk_sys::GtkFlowBoxChildClass;
+            let f = (*parent_class)
+                .activate
+                .expect("No parent class impl for \"activate\"");
```
Comment (sdroege) [inline gtk4/src/subclass/flow_box_child.rs:29]: Optional

File: gtk4/src/subclass/list_box_row.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+impl<T: ListBoxRowImpl> ListBoxRowImplExt for T {
+    fn parent_activate(&self, row: &Self::Type) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gtk_sys::GtkListBoxRowClass;
+            let f = (*parent_class)
+                .activate
+                .expect("No parent class impl for \"activate\"");
```
Comment (sdroege) [inline gtk4/src/subclass/list_box_row.rs:28]: Optional

**Review by sdroege — COMMENTED** [review]
Looks good to me otherwise. Please check all these against gtk3 too so that we keep them consistent.

**Comment by bilelmoussaoui** [comment]
> Looks good to me otherwise. Please check all these against gtk3 too so that we keep them consistent.

GtkListBoxRow is similar to the gtk3 one, GtkListBox had all of it's vfunc removed & the rest are missing from gtk3. I don't think there are many we can copy from gtk3 other than the Cell* ones

File: gtk4/src/subclass/drawing_area.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use glib::Cast;
+
+use super::widget::WidgetImpl;
+use DrawingArea;
+use Widget;
+
+pub trait DrawingAreaImpl: DrawingAreaImplExt + WidgetImpl {
+    fn resize(&self, drawing_area: &Self::Type, width: i32, height: i32) {
```
Comment (sdroege) [inline gtk4/src/subclass/drawing_area.rs:12]: Not sure why it has a vfunc for resizing but nothing for drawing... oh well! :)

Comment (bilelmoussaoui) [reply gtk4/src/subclass/drawing_area.rs:12]: The drawing happens on snapshot at widget level iirc.

Comment (bilelmoussaoui) [reply gtk4/src/subclass/drawing_area.rs:12]: & the resize was added as people always misused size-allocate for drawing area, thankfully that was removed

**Pull request merged**